sql-plugin/src/main/scala/com/nvidia/spark/rapids/CostBasedOptimizer.scala

```diff
@@ -115,25 +118,50 @@ class CostBasedOptimizer(conf: RapidsConf) extends Logging {
     if (totalGpuCost > totalCpuCost) {
       // we have reached a point where we have transitioned onto GPU for part of this
       // plan but with no benefit from doing so, so we want to undo this and go back to CPU
-      if (plan.canThisBeReplaced) {
+      if (plan.canThisBeReplaced && !consumesQueryStage(plan)) {
         // this plan would have been on GPU so we move it and onto CPU and recurse down
         // until we reach a part of the plan that is already on CPU and then stop
         optimizations.append(ReplaceSection(plan, totalCpuCost, totalGpuCost))
-        plan.recursiveCostPreventsRunningOnGpu()
+        recursiveCostPreventsRunningOnGpu(plan)
       }
 
       // reset the costs because this section of the plan was not moved to GPU
       totalGpuCost = totalCpuCost
     }
 
-    if (!plan.canThisBeReplaced) {
+    if (!plan.canThisBeReplaced || consumesQueryStage(plan)) {
       // reset the costs because this section of the plan was not moved to GPU
       totalGpuCost = totalCpuCost
     }
 
     (totalCpuCost, totalGpuCost)
   }
 
+  /**
+   * Recursively force a section of the plan back onto CPU, stopping once a plan
+   * is reached that is already on CPU or cannot be moved back onto CPU.
+   */
+  private def recursiveCostPreventsRunningOnGpu(plan: SparkPlanMeta[_]): Unit = {
+    if (plan.canThisBeReplaced && !consumesQueryStage(plan)) {
+      plan.costPreventsRunningOnGpu()
+      plan.childPlans.foreach(recursiveCostPreventsRunningOnGpu)
```

Comment:
I am starting to get confused with the algorithm and what it really intends to do so I have a hard time judging what is and isn't a bug.  Because now that we stop when we hit a consumsQueryStage doesn't that mess up the cost calculations that we did to get here?

Could we have some high level comments  about what the algorithm is trying to do?

Reply:
I've added some more documentation but the intent is that if we see an operator that cannot run on GPU (because we don't support it) or an operator that is consuming a query stage (so has to stay wherever it is) then we make the GPU cost be the same as the CPU cost. We only have cost differences where this actually potential to change the plan. At least, that is the intent.